core/src/main/java/ai/timefold/solver/core/impl/solver/termination/AdaptiveTermination.java

```diff
@@ -0,0 +1,171 @@
+package ai.timefold.solver.core.impl.solver.termination;
+
+import java.util.NavigableMap;
+import java.util.TreeMap;
+
+import ai.timefold.solver.core.api.score.Score;
+import ai.timefold.solver.core.impl.phase.scope.AbstractPhaseScope;
+import ai.timefold.solver.core.impl.phase.scope.AbstractStepScope;
+import ai.timefold.solver.core.impl.solver.scope.SolverScope;
+import ai.timefold.solver.core.impl.solver.thread.ChildThreadType;
+
+import org.jspecify.annotations.NonNull;
+import org.jspecify.annotations.Nullable;
+
+public final class AdaptiveTermination<Solution_, Score_ extends Score<Score_>> implements Termination<Solution_> {
+    private final long gracePeriodMillis;
+    private final double minimumImprovementRatio;
+
+    boolean isGracePeriodActive;
+    private long gracePeriodStartTimeMillis;
+    private LevelScoreDiff gracePeriodScoreDiff;
+
+    private final NavigableMap<Long, Score_> scoresByTime;
+
+    public AdaptiveTermination(long gracePeriodMillis, double minimumImprovementRatio) {
+        this.gracePeriodMillis = gracePeriodMillis;
+        this.minimumImprovementRatio = minimumImprovementRatio;
+        this.scoresByTime = new TreeMap<>();
```

Comment:
I was skeptical of this approach, as fast-stepping algorithms will easily put tens of thousands of entries to this map per second, and I was expecting that to be both CPU-intensive and memory-intensive.

However, in the experiments I ran, this did not materialize. Impact of the termination is barely visible in the data. Interesting.

Reply:
I wonder though if the `TreeMap` is bringing any tangible benefits. Consider this:

- We don't actually need sorting, do we? The values are sequential, a `List` would do.
- We are forced to box `long` into `Long`, causing some unnecessary GC pressure.

I wonder if we could convert this into `List<Score_>`. The index would have to become relative to the last reset, so starting with zero.